clang/test/SemaSYCL/sycl-restrict.cpp

```diff
@@ -0,0 +1,24 @@
+// RUN: %clang_cc1 -triple spir64 -fsycl-is-device -verify -fsyntax-only %s
+
+void eh_ok(void)
+{
+  __float128 A;
+}
+
+void usage() {
+  // expected-error@+1 {{__float128 is not supported on this target}}
+  __float128 A;
+}
+
```

Comment:
I have a couple more concerns with this... we would obviously like __float128 to be usable in unevaluated contexts, so can we test that? (Basically in a decltype/sizeof/etc).    I'd also like to see some more indirection through template cases (at least eventually).

Reply:
Okay, I've made some testing.
I see the following problems:
- Using of __float128 is not diagnosed in function return value and parameters, but It should, because it add unsupported type to resulting IR i.e. 
```
__float128 foo(__float128 P) { // is not diagnosed if this function is called from device code
...
}
```
- Using decltype with underlying __float128 is not diagnosed but It should, because it add unsupported type to resulting IR:
```
__float128 foo1() { // is not diagnosed if this function is called from device code
...
}
decltype(foo1()) D; // is not diagnosed if used in device code
```
- Using of __float128 is diagnosed in sizeof operator, which, I suppose, shouldn't, because It doesn't adds unsupported type to resulting IR:
```
int E = sizeof(__float128); // diagnosed in device code
```
- Is not diagnosed if used in classes:
```
template <class T>
class Z {
public:
  T field;
  __float128 field1;
};
```
But if the last problem was expected by me, because we can't defer diagnostics which comes from non-function lexical contexts. Other ones were a surprise. Not sure where it comes from yet - problem with deferred diagnostics or place for diagnosing in Sema is wrong.

Reply:
> 
> 
> Okay, I've made some testing.
> I see the following problems:
> 
>     * Using of __float128 is not diagnosed in function return value and parameters, but It should, because it add unsupported type to resulting IR i.e.
> 
> 
> ```
> __float128 foo(__float128 P) { // is not diagnosed if this function is called from device code
> ...
> }
> ```

Yeah, that is definitely a problem.

>     * Using decltype with underlying __float128 is not diagnosed but It should, because it add unsupported type to resulting IR:
> 
> 
> ```
> __float128 foo1() { // is not diagnosed if this function is called from device code
> ...
> }
> decltype(foo1()) D; // is not diagnosed if used in device code
> ```

Both of those SHOULD diagnose, but you can do some stuff with decltype and metaprogramming that doesn't result in it making it to IR.  


>     * Using of __float128 is diagnosed in sizeof operator, which, I suppose, shouldn't, because It doesn't adds unsupported type to resulting IR:
> 
> 
> ```
> int E = sizeof(__float128); // diagnosed in device code
> ```

Yep, exactly.  Though again, this isn't sizeof specific, decltype doesn't have to introduce the type either.  Consider a case where it is a template parameter, but never otherwise instantiated:

    template<typename t> void foo(){};
    foo<__float128>(); 

Other than mangling, the float128 does nothing there.



>     * Is not diagnosed if used in classes:
> 
> 
> ```
> template <class T>
> class Z {
> public:
>   T field;
>   __float128 field1;
> };
> ```
> 
> But if the last problem was expected by me, because we can't defer diagnostics which comes from non-function lexical contexts. Other ones were a surprise. Not sure where it comes from yet - problem with deferred diagnostics or place for diagnosing in Sema is wrong.

Reply:
```
template<typename t> void foo(){};
foo<__float128>(); 
```
This is diagnosed. I guess, this is a problem too.
Thanks for the test case.

> underlying __float128 is not diagnosed but It should, because it add unsupported type to resulting IR:
__float128 foo1() { // is not diagnosed if this function is called from device code
...
}
decltype(foo1()) D; // is not diagnosed if used in device code
Both of those SHOULD diagnose, but you can do some stuff with decltype and metaprogramming that doesn't result in it making it to IR.

Not sure that I understand. Could you please elaborate a bit more?

Reply:
Decltype can be used in a bunch of places where it doesn't actually introduce an IR type, I think that is all I was getting at.  Even something like the foo<__int128> case above is just a restatement of this case.

Consider something like:

    __float128 nonemittedfunc();
    std::conditional_t<SomeI < 1, decltype(nonemittedfunc()), int> SomeVar;

It is similar to the case, just saying that decltype doesn't necessarily mean that the type is introduced.

Reply:
Okay, I extended the test and made some analysis, function parameters and return values is not diagnosed because in this case `getCurLexicalContext()` returns translation unit decl and we can diagnose only
inside function decls.
I was able to suppress diagnosing of `sizeof` expressions by adding a new check in `SYCLDiagIfDeviceCode` on whether we are in unevaluated context.
Didn't analyze decltype problems yet...